Add one-time login flow

emcsween opened the pull request:
### Description

* The user makes a one-time login request by providing their email
* An email containing a link to a one-time login endpoint is sent to the
  user
* The endpoint logs the user in and discards the token so that the link
  cannot be used repeatedly

#### Screenshots

![one-time-login-1](https://user-images.githubusercontent.com/5454374/60398051-ef264700-9b21-11e9-917b-0ee6721c43b3.png)
![one-time-login-2](https://user-images.githubusercontent.com/5454374/60398052-ef264700-9b21-11e9-9e65-61fb68f04f93.png)
![one-time-login-3](https://user-images.githubusercontent.com/5454374/60398053-efbedd80-9b21-11e9-88b7-d643799aa651.png)
![one-time-login-4](https://user-images.githubusercontent.com/5454374/60398054-efbedd80-9b21-11e9-8546-013227a16009.png)

#### Related Issues / PRs

* Issue: https://github.com/overleaf/google-ops/issues/334
* dev-environment PR: https://github.com/overleaf/dev-environment/pull/232

### Review

There is quite a bit of messaging involved. I'll need someone to review the copy.

I added a logout endpoint to ease testing, but I didn't include it in the UI yet. This is probably something we will want to add in another PR.

#### Potential Impact

This mechanism will be used by users who don't have a password associated with their Overleaf accounts. This includes people who authenticate via Google, Twitter, etc.

#### Manual Testing Performed

- [x] Follow the one-time login flow
- [x] Try to use a bad token
- [x] Login with password
- [x] Try to login with a bad password

#### Accessibility

Every form field is easily accessible via tabbing.

### Deployment

The read-only site is not deployed yet.

emcsween committed:
Add one-time login flow

* The user makes a one-time login request by providing their email
* An email containing a link to a one-time login endpoint is sent to the
  user
* The endpoint logs the user in and discards the token so that the link
  cannot be used repeatedly

emcsween committed:
Add missing alert styles

The alert styles were lost in the rebranding from ShareLaTeX to
Overleaf.

briangough commented on app.coffee:
```diff
@@ -1,37 +1,51 @@
 app.set('view engine', 'pug')
 
 app.use(SessionMiddleware.middleware)
+app.use(bodyParser.urlencoded({ extended: false }))
```
One for another PR but we need to rate-limiting middleware too.

briangough commented on app/coffee/AuthController.coffee:
```diff
@@ -0,0 +1,57 @@
+				else
+					next(err)
+				return
+			logger.info({ email, userId }, "successful login")
```
how about
```
if err
   ....
else
   logger.info({ email, userId }, "successful login")
   ....
```
to avoid failling through to the success case.

I've always written `if err?`  - seems safer as it will catch any error value.

emcsween replied:
Me too, but I stopped doing it following this discussion on overleaf-dev: https://digital-science.slack.com/archives/C20TZCMMF/p1559316110035600

I'm not sure who to follow at this point ;)

briangough replied:
Ah yes I remember that conversation now, maybe the `?` doesn't matter but I think it's good to have an `else` for the success case.

briangough commented on app/coffee/AuthController.coffee:
```diff
@@ -0,0 +1,57 @@
+				else
+					next(err)
+				return
+			logger.info({ email, userId }, "successful one-time login")
```
comment about else as above

briangough commented on app/coffee/AuthController.coffee:
```diff
@@ -0,0 +1,57 @@
+
+	oneTimeLoginRequest: (req, res, next) ->
+		{ email } = req.body
+		AuthHandler.generateOneTimeLoginToken email, (err, token) ->
```
I think we should check that the email address exists for a user in the database  before sending out a token, otherwise we allow sending to arbitrary email addresses which could create problems with spamming.

emcsween replied:
very good point :+1:

briangough commented on app/coffee/AuthHandler.coffee:
```diff
@@ -0,0 +1,64 @@
+		update = {
+			$set: { usedAt: now }
+		}
+		db.tokens.findAndModify { query: query, update: update }, (err, token) ->
```
`tokens` needs to be added to the list of allowed collections in `mongojs.coffee`

How about we give it a unique name like `readOnlyLoginTokens` so it does not clash with the `tokens` collection already used by `web` here [mongojs.js#L8-L16](https://github.com/overleaf/web-internal/blob/ed8636630c1debcb21cea16e50adddcb636631af/app/src/infrastructure/mongojs.js#L8-L16)?

emcsween replied:
I'm good to make the collection name more explicit, like `oneTimeLoginTokens`, but it's not necessary to indicate which app the collection belongs to because the collections are in a separate database.

Also, read-only has `users` and `projects` collections that are different from the corresponding collections in web.

briangough replied:
Ah, I hadn't noticed it was in a separate `read-only` database at [settings.defaults.coffee#L33](https://github.com/overleaf/read-only/blob/36e962fdb8dc790de2c0ddab27c3014d93d27038/config/settings.defaults.coffee#L33) in the dev environment.

briangough commented on app/coffee/AuthHandler.coffee:
```diff
@@ -0,0 +1,64 @@
+				return callback(err)
+			if !token
+				return callback(new Errors.AuthenticationError("invalid token"))
+			db.users.findOne { email: normalizedEmail }, (err, user) ->
```
A user can have multiple email addresses for an account, we're only checking the primary address here (I think that's ok but maybe there could be confusion if they are used to logging in with another address?).

jdleesmiller replied:
Correct... probably one day we will allow login with secondary emails on the main site, but at the moment it's the same there, so hopefully it won't be too surprising.

briangough commented on app/coffee/AuthHandler.coffee:
```diff
@@ -0,0 +1,64 @@
+			createdAt: now
+			expiresAt: expiresAt
+		}
+		db.tokens.insert doc, (err) ->
```
We're allowing multiple tokens to be inserted here (unless there's a unique index on `email`), only limited by rate-limiting.  How about we do an mongo `update` with the `upsert:true` option so that old tokens are no longer valid when a new one is generated?

jdleesmiller replied:
In the past we've seen problems with this kind of setup, because sometimes there's a delay in getting the emails, so people fill in the form 5 times and then get 5 emails in random order, only one of which will actually work. If they're relatively short lived tokens and there is a rate limiter on issuing them, perhaps it is OK to allow multiple tokens?

emcsween replied:
ok. I'll leave it as is.

briangough commented on app/coffee/AuthHandler.coffee:
```diff
@@ -0,0 +1,64 @@
+	oneTimeLogin: (email, token, callback) ->
+		now = new Date()
+		normalizedEmail = email.toLowerCase()
+		query = {
```
I think it would be good to put somewhere in this file what indexes we are going to use. (e.g. [MongoManager.coffee#L82-L97](https://github.com/overleaf/track-changes/blob/209835b36d926afdb793ce6e5625e918e9be1660/app/coffee/MongoManager.coffee#L82-L97) or [UserManager.coffee#L5-L9](https://github.com/overleaf/github-sync/blob/397df2123f8b0b850783fbb225618006dfe46f1f/app/coffee/UserManager.coffee#L5-L9))

In particular, we should have a TTL index on `expiresAt` to avoid the collection growing forever.

briangough commented on app/coffee/AuthHandler.coffee:
```diff
@@ -0,0 +1,64 @@
+		logger.info({ email }, "Generating one-time login token")
+		now = new Date()
+		expiresAt = new Date(now.getTime() + TOKEN_TTL_MS)
+		token = crypto.randomBytes(32).toString('hex')
```
This is optional but I've always thought we should be hashing our tokens rather than storing them in plain text, because they are like passwords. So how about we split the token into two 16-byte parts and store one part as plain text and the other hashed with bcrypt?  We use the plain text comparison first so that the expensive bcrypt comparison is not open to everyone.

This would also  prevent timing attacks on the comparison of token with the stored version.

jdleesmiller replied:
I think bcrypt may be over the top here, because these are long and random tokens (as opposed to user passwords that may be short and reused on other sites) protected by a rate limiter, and they are short lived (these tokens should only last an hour or two).

I think timing attacks are a concern, however, because we are looking up the whole token in the database (together with the email address). An approach we have taken elsewhere is to use a database-generated ID and a random token together. The ID part guarantees uniqueness, and you can use it for the lookup; then you can then do a timing-safe compare on the token (see for example how we do it [in web](https://github.com/overleaf/web/blob/1c30916030261d8bb295e28b10cd632033b80cb3/app/src/Features/TokenAccess/TokenAccessHandler.js#L98-L101)).

(If we only allowed one token per email address, we could look up on the email only and then do a timing-safe compare on the token, but as mentioned above, I think limiting it to one token per email address may not be a very good UX.)

emcsween replied:
Thanks for the timing considerations. I hadn't thought of that. Would it work if I combine your two approaches?

* Split the token in two
* Use the first part of the token in the db query
* Compare the whole token using `crypto.timingSafeEqual()`

If I understand correctly, this would limit any timing attack to discovering only the first half of the token.

jdleesmiller replied:
I think that would work, yes. If the tokens already have an ID from Mongo, I think using it would make sense... but it would require some changes downstream to pass both parameters. In the presence of a fairly low rate limiter, I think it's probably OK as-is.

briangough commented on app/coffee/AuthHandler.coffee:
```diff
@@ -0,0 +1,64 @@
+				if err
+					return callback(err)
+				if !match
+					return callback(new Errors.AuthenticationError("incorrect password"))
```
Maybe one for another PR but we should record a failed login count and lock the account if there are too many failures.

briangough commented on app/coffee/AuthHandler.coffee:
```diff
@@ -0,0 +1,64 @@
+			if err
+				return callback(err)
+			if !token
+				return callback(new Errors.AuthenticationError("invalid token"))
```
As above for locking the account if there are too many failures.

briangough commented on app.coffee:
```diff
@@ -1,37 +1,51 @@
 app.set('views', __dirname + '/app/views')
 app.set('view engine', 'pug')
 
 app.use(SessionMiddleware.middleware)
```
One for another PR but we should make sure that the sessions have a short lifetime and are expired from the database as it is likely the user will not "log out" of the readonly site.

emcsween replied:
I made them expire after 6 hours. Trying to find a balance between convenience and security. Let me know what you think.

briangough reviewed (changes requested):
Changes:

- add collection to mongojs.coffee, use unique name
- check email address exists in user collection before sending one-time token email
- document indexes/TTLs
- query re multiple vs single valid tokens

Others are all comments or notes for other PRs.

jdleesmiller commented on app/coffee/EmailHandler.coffee:
```diff
@@ -0,0 +1,37 @@
+The link below will allow you to log in once to the Overleaf maintenance site.
+
+Once logged in, you'll be able to download a read-only copy of your projects
+while we bring the site back.
```
The copy is good already, but for your consideration:

> The link below will log you into Overleaf's read only maintenance site, so you can download your projects.

> The link is valid for one hour or until the maintenance is finished.

(This suggests an optional 'nice touch' in the main application: also add this route to web and put up a page (ignoring the token) explaining that the site is back and they can log in normally.)

jdleesmiller commented on app/coffee/EmailHandler.coffee:
```diff
@@ -0,0 +1,37 @@
+Once logged in, you'll be able to download a read-only copy of your projects
+while we bring the site back.
+"""
+			ctaText: "Log in to Overleaf"
```
Maybe "Access my Projects"? That is probably closer to the user's end goal.

jdleesmiller commented on app/coffee/EmailHandler.coffee:
```diff
@@ -0,0 +1,37 @@
+	sendOneTimeLoginEmail: (email, token, callback) ->
+		logger.info({ email }, "Sending one-time login message")
+		templateVars = {
+			title: "Overleaf login"
```
Maybe "Overleaf Read Only Access"?

jdleesmiller commented on app/coffee/EmailHandler.coffee:
```diff
@@ -0,0 +1,37 @@
+		mailerParams = {
+			to: email
+			from: Settings.email.fromAddress
+			subject: "Overleaf login"
```
Can change to match the title, if we change the title as suggested above.

jdleesmiller commented on app/coffee/AuthController.coffee:
```diff
@@ -0,0 +1,57 @@
+	login: (req, res, next) ->
+		{ email, password } = req.body
+		logger.info({ email }, "received password login request")
+		email = email.trim()
```
~Should we also `toLowerCase()` here? I believe that is what web does.~ Ah, I see it is done later. Also fine.

Was going to link to this helper; it may be useful to extract a helper here, too, for consistency: https://github.com/overleaf/web/blob/1c30916030261d8bb295e28b10cd632033b80cb3/app/src/Features/Helpers/EmailHelper.js

emcsween replied:
Nice one! I have an item for validation in my improvements list. I'll keep this in mind.

jdleesmiller commented on app/views/one-time-login-request-form.pug:
```diff
@@ -0,0 +1,24 @@
+			.col-md-6.col-md-offset-3
+				.card
+					p
+						| Please enter your email below and we'll send you a
```
Comma before `and`. And to echo the wording, maybe

> ... link that will log you into Overleaf's read only maintenance site, so you can download your projects.

jdleesmiller commented on app/views/home.pug:
```diff
@@ -7,8 +7,10 @@ block content
+						if failedLogin
 							.alert.alert-danger Your email or password was incorrect
+						if failedOneTimeLogin
+							.alert.alert-danger Your login link has expired
```
I think it's always a good thing to advise the user what to do next when there's an error. So maybe "Your login link has expired. Please request another one." (ideally linking back to the place they can do that).

jdleesmiller commented on app/views/one-time-login-email-sent.pug:
```diff
@@ -0,0 +1,12 @@
+			.col-md-6.col-md-offset-3
+				.card
+					p
+						| We have sent you an email with a link that will allow
```
To echo the wording I suggested in the email, how about:

> We have sent you an email with a link that will log you into Overleaf's read only maintenance site, so you can download your projects. Please check your inbox and spam folder.

jdleesmiller reviewed (commented):
Looks good. I've had a look at the wording as requested and made some suggestions (but it's not bad as-is). And I commented on some tech stuff.

emcsween commented:
Thanks for the review. I have made changes in subsequent commits.

jdleesmiller commented on app/coffee/AuthHandler.coffee:
```diff
@@ -0,0 +1,85 @@
+			usedAt: { $exists: false }
+		}
+		update = {
+			$set: { usedAt: now }
```
This will optimistically set `usedAt` even if the comparison on the latter half of the token fails. That's probably OK, but it is a bit odd. Also, what happens if you pass an empty token?

emcsween replied:
Right. I'll do it in two steps.

If you pass an empty token, it will get you one of your tokens and fail at the comparison step. I'll add a test for it.

jdleesmiller commented on app/coffee/AuthHandler.coffee:
```diff
@@ -0,0 +1,85 @@
+	_compareTokens: (token1, token2) ->
+		token1Buffer = Buffer.from(token1)
+		token2Buffer = Buffer.from(token2)
+		if token1Buffer.length != token2Buffer.length
```
Maybe it would be simpler to just check the total length of the token as a validation step, before we even query for it?

emcsween replied:
I was thinking that it's easier if the token checking part doesn't assume anything about the length of the token. If, for example, we change the length of the generated token for some reason, the token check works with both versions during the transition.

jdleesmiller replied:
OK, good point. That is a good property to have.

jdleesmiller commented on app/coffee/SessionMiddleware.coffee:
```diff
@@ -2,9 +2,12 @@ Settings = require("settings-sharelatex")
 session = require("express-session")
 MongoStore = require('connect-mongo')(session)
 
+SESSION_TTL_SECS = 6 * 60 * 60   # 6 hours
```
I think we could probably go lower on this --- people won't be using this site for very long, hopefully, to grab their projects. One hour?

jdleesmiller commented on app/coffee/AuthHandler.coffee:
```diff
@@ -0,0 +1,85 @@
+Errors = require("./Errors")
+
+TOKEN_TTL_MS = 60 * 60 * 1000    # 1 hour
+TOKEN_SEARCH_PREFIX_LENGTH = 16  # 8 bytes
```
The collision probability seems [low enough](http://davidjohnstone.net/pages/hash-collision-probability), so this is probably long enough.

emcsween replied:
:) yes. and this is only among tokens for a specific email, of which there is reasonably only a handful.

briangough commented on app/coffee/AuthHandler.coffee:
```diff
@@ -0,0 +1,93 @@
+		normalizedEmail = email.toLowerCase()
+		async.auto {
+			tokenDoc: (cb) =>
+				tokenPrefix = token.slice(0, TOKEN_SEARCH_PREFIX_LENGTH)
```
to avoid any unexpected behaviour with arbitrary user input for the token being used to create a Regexp, how about we assume that the token will always be hex digits `[0-9a-f]` (or alphanumeric `[0-9a-z]`) and validate it against that first?

emcsween replied:
I agree, but I think this validation goes at the controller level. I had this as a nice-to-have next PR. I now realize that maybe I should add it right away. I was thinking of introducing a validation library instead of doing it ad hoc.

At this level, we are protected from arbitrary regexps by `_.escapeRegExp()` in the next line.

briangough reviewed (changes requested):
I think we should have some validation on the token before constructing a regex on it.

I noticed we don't have any csrf protection on these forms, but as the site is by its nature read-only and doesn't do anything to the users account maybe this doesn't matter.

jdleesmiller commented:
I also wondered about CSRF protection, but I think you are right that for a read only site it is not strictly speaking required. It is probably worth doing for completeness, but that can come later.

emcsween commented:
I added validation in the last commit. I also split the router into its own module because the main file was becoming a bit big.

I didn't make an effort to offer good validation messages since those are only login forms. Basically, validation errors will appear as "login failed" messages.
